### ISSUE TITLE
DCAT-US Harvester

### ISSUE THREAD

**Issue by kimwdavidson** [GSA/datagov-ckan-multi#301]
As a federal agency's data publisher, I want to describe my data in one place using federal standards and see it published on data.gov, so that I can comply with federal mandates in the fewest steps.

## Acceptance criteria
- [x] GIVEN I add a GSA data.json harvest source
  WHEN I trigger a harvest run of that source
  THEN the harvest is successful and the metadata is ingested and visible on the [GSA organization page](https://catalog-next.sandbox.datagov.us/organization/gsa-gov)
  ~AND a message is delivered to SMTP server (viz. we don't send an email but the scaffolding is in place so that an email could be sent)~ (captured in https://github.com/GSA/datagov-ckan-multi/issues/315) 

## Details/Tasks
- [x] Make upstream harvester extension a dependency to catalog next CKAN 2.8
- [x] Make the data.json (aka DCAT US) ext a dependency to catalog next CKAN 2.8
- [x] Manual tests that this works for different sources
- [x] Import harvester sources which are covered by this last #302  
- [x] PR approved by Aaron
- [x] Scaffolding for email is in place (message sent to SMTP server)
   - [x] Create a custom extension in catalog datagov extension and port over the Harvester code that customized the previous fork of the Harvester extension [this has the code for email notifications so we can re-use it]
- [x] Review the current tests. If we think there should be more then we should add a ticket for this. 
- [ ] PR approved by Aaron 

Next steps (brain dump):
- [x] List of data.sources with RAG status (red = failling, amber = bugs, green = running fine). There are 150+ of these. Covered on [#312](https://github.com/GSA/datagov-ckan-multi/issues/312)
- [x] GSA to advise which of these need to be green (ie are there any sources that were working before but now are not and id any sources that had issues before that we could fix) 
- [ ] Deploy to sandbox/staging the agreed DCATUS harvester source scope and QA the work

**Comment by avdata99** [comment]
Manual tests:
 - Library of Congress  source: https://www.loc.gov/static/json/data.json. **Works fine**
  - City of Boise harvest source: http://opendata.cityofboise.org/data.json: **Fail** (requeire analysis on non-federal 
validation

**Comment by avdata99** [comment]
Testing sources from list @hkdctol in my local environment 

From CFPB JSON | cfpb-json at http://www.consumerfinance.gov/data.json

![image](https://user-images.githubusercontent.com/3237309/82582781-eb037d80-9b68-11ea-819f-fbb05cc4c3fa.png)

![image](https://user-images.githubusercontent.com/3237309/82582810-f9519980-9b68-11ea-9d89-690002cc66d4.png)

![image](https://user-images.githubusercontent.com/3237309/82582855-0e2e2d00-9b69-11ea-86c0-67aaf3c6ef67.png)

![image](https://user-images.githubusercontent.com/3237309/82582886-1c7c4900-9b69-11ea-8cf3-6a74953e23bd.png)

**Comment by avdata99** [comment]
California: 499 errors; 1 OK

![image](https://user-images.githubusercontent.com/3237309/82584068-5437c080-9b6a-11ea-818d-e64ba2eec2a8.png)

Sample error
```
Identifier: 0d6af8c7-bdcb-4206-a222-fea8593d42c1; Title: CHDP Summary Of All Health Assessments By Funding Source SFY 2008-2014; 1 Error(s) Found. ### ERROR #1: 'distribution':[{'downloadURL': 'https://data.chhs.ca.gov/dataset/886e17f7-cd4e-411a-99fc-2d53925e6075/resource/4f35a8fd-38f1-43e3-98c2-41a524147f54/download/chdp-summary-of-all-health-assessments-by-funding-source-sfy-2008-2014.csv', 'title': 'CHDP Summary Of All Health Assessments By Funding Source SFY 2008-2014 (CSV)', 'format': 'CSV', '@type': 'dcat:Distribution', 'description': 'The Child Health and Disability Prevention (CHDP) annual report for state fiscal years (SFY) specified in the title provides an aggregate level data drawn from the Confidential Screening/Billing Report (PM-160) form used for fee-for-service (FFS) reimbursement of services, and/or for reporting preventive health assessment services. The display of the data in this report is intended to provide a detailed view of the health assessments provided to CHDP eligible children and youth as reported to the California Department of Health Care Services (DHCS) during SFY 2008-12. CHDP eligible children and youth are those whose gross family incomes are at or below 200 percent of the federal poverty level, age restriction for full-scope Medi-Cal of less than 21 years of age, presumptive eligibility (the Gateway process), or children and youth with no health insurance for health assessments who must be less than 19 years of age. CHDP children and youth receive health assessments and screenings according to the age-based CHDP Periodicity Schedule or when assessments and screenings are considered medically necessary. Note: Empty Cells represent values that are not shown to protect confidentiality of the individuals summarized in the data.'}, {'downloadURL': 'https://data.chhs.ca.gov/dataset/886e17f7-cd4e-411a-99fc-2d53925e6075/resource/a4e737e3-2ddd-4fd6-952a-217b64eb244a/download/himdopen-datadata-dictionary-chdp-summary-of-all-health-assessments-sfy-2008-12.docx', 'title': 'data-dictionary-chdp-summary-of-all-health-assessments-sfy-2008-12', 'format': 'DOCX', '@type': 'dcat:Distribution', 'description': 'The Child Health and Disability Prevention (CHDP) annual report for state fiscal years (SFY) specified in the title provides an aggregate level data drawn from the Confidential Screening/Billing Report (PM-160) form used for fee-for-service (FFS) reimbursement of services, and/or for reporting preventive health assessment services. The display of the data in this report is intended to provide a detailed view of the health assessments provided to CHDP eligible children and youth as reported to the California Department of Health Care Services (DHCS) during SFY 2008-12. CHDP eligible children and youth are those whose gross family incomes are at or below 200 percent of the federal poverty level, age restriction for full-scope Medi-Cal of less than 21 years of age, presumptive eligibility (the Gateway process), or children and youth with no health insurance for health assessments who must be less than 19 years of age. CHDP children and youth receive health assessments and screenings according to the age-based CHDP Periodicity Schedule or when assessments and screenings are considered medically necessary. Note: Empty Cells represent values that are not shown to protect confidentiality of the individuals summarized in the data.'}, {'downloadURL': 'https://data.chhs.ca.gov/dataset/886e17f7-cd4e-411a-99fc-2d53925e6075/resource/ca40902f-673a-48d2-8ad1-81eeff4f9d95/download/chdp-summary-of-all-health-assessments-by-funding-source-sfy-2008-2014-t6pt0i.zip', 'title': 'All resource data', '@type': 'dcat:Distribution', 'format': 'ZIP'}] is not valid under any of the given schemas.
```

**Comment by estebanruseler** [comment]
Shared with Hyon on Slack Wednesday (5/20) https://docs.google.com/spreadsheets/d/1LgVvgiUC5Vc-OQaSmtvQNlTFLhs3Mho65ekggYEiYh4/edit?ts=5e4177f8#gid=264788117

**Comment by avdata99** [comment]
Final analysis and steps to close this issue
   - We did some manual tests on harvesting DCAT-US sources. 
   - We create a new ext to include GSA customization in the future. 
       - [ ] Should we maintain this new ext in the PR?
   - We created a [list of all (150) DCAT-US sources](https://docs.google.com/spreadsheets/d/1LgVvgiUC5Vc-OQaSmtvQNlTFLhs3Mho65ekggYEiYh4/edit?ts=5e4177f8#gid=264788117) from current catalog.data.gov.
       - [ ] We need a new issue to analyze the 150 DCAT-US harvest sources. This issue should allow us to create new issues to fix specific harvesting DCAT-US errors. Test covering this error should be added when possible.
   - [ ] Create a new issue at upstream harvest ext to allow more notifications.
   - [ ] We suggest we add an issue to increase test coverage in the backlog for future priortization.
  - Do we need a way to full test extensions inside out env? We detect an error to run that tests (`reset_db` function freezes the tests)
      - [ ] Create an issue to fix the env to run the extensions test
          - [ ] Run DCAT-US extension tests: fails at `test_datajsonharvester.py", line 42, in setup: reset_db()` ERROR: `ProgrammingError: (psycopg2.ProgrammingError) relation "package_tag_revision" does not exist LINE 1: delete from "package_tag_revision"`
          - [ ] To test notification/email in harvest upstream ext inside this env: Fail when we try to use `ckanext.harvest.tests.nose.test_action:TestHarvestErrorMail`

**Comment by kimwdavidson** [comment]
Blocked because we can't test in the sandbox

**Comment by kimwdavidson** [comment]
Blocked. Ready for QA when sandbox is ready.

**Comment by avdata99** [comment]
**QA**

Harvest works at the sandbox

![image](https://user-images.githubusercontent.com/3237309/84061562-7fa51280-a994-11ea-95de-18aef9542bd2.png)

**We can see datasets**

![image](https://user-images.githubusercontent.com/3237309/84061685-c135bd80-a994-11ea-9f49-08cb4193a042.png)

**Comment by adborden** [comment]
I was unable to create a new DCAT-US harvest source and ran into https://github.com/GSA/datagov-ckan-multi/issues/321

**Comment by adborden** [comment]
We discussed the upstream behavior of ckanext-harvester vs our fork. The Harvest Sources tab and the controller serving that is a data.gov customization. For now, this behavior isn't essential to QA, so we'll go ahead.

**Comment by avdata99** [comment]
> We discussed the upstream behaviour of ckanext-harvester vs our fork. The Harvest Sources tab and the controller serving that is a data.gov customization. For now, this behavior isn't essential to QA, so we'll go ahead.

If you are able to QA using the `/harvest` URL maybe we can finish QA and close this issue.
We have a [PR](https://github.com/GSA/ckanext-datagovtheme/pull/44) to fix #321. We can leave this for future analysis and move 321 to the _backlog_ column.

**Comment by adborden** [comment]
This is looking good enough.

I was able to create the GSA org and a [DCAT-US (data.json) harvest source](https://catalog-next.sandbox.datagov.us/harvest/admin/gsa-data-json).

The job started successfully, but after 24 hours it is still in the running state. I opened https://github.com/GSA/datagov-ckan-multi/issues/324 to track this.

I inspected the harvest logs and verified the job moved from gather to fetch queues where it got stuck.

https://catalog-next.sandbox.datagov.us/organization/general-services-administration shows some datasets.

Because the job never successfully completed, I could not verify the postfix logs, but I feel that AC is captured in other stories.
